include/xtensor/xreducer.hpp

```diff
@@ -128,36 +128,36 @@ namespace xt
         }
 
         // TODO remove all entries where iter_shape == 0
-        iter_shape.erase(iter_shape.begin() + last_ax, iter_shape.end());
-        iter_strides.erase(iter_strides.begin() + last_ax, iter_strides.end());
+        iter_shape.erase(iter_shape.begin() + (ptrdiff_t) last_ax, iter_shape.end());
+        iter_strides.erase(iter_strides.begin() + (ptrdiff_t) last_ax, iter_strides.end());
 
         for (std::size_t i = 0; i < iter_shape.size(); ++i)
         {
             if (iter_shape[i] == 0)
             {
-                iter_shape.erase(iter_shape.begin() + i);
-                iter_strides.erase(iter_strides.begin() + i);
+                iter_shape.erase(iter_shape.begin() + (ptrdiff_t) i);
+                iter_strides.erase(iter_strides.begin() + (ptrdiff_t) i);
                 ++i;
             }
         }
```

Comment:
There's still a problem though; let's say you erase `iter_shape.begin()` (because `i = 0), when starting next iteration, `i = 2`, you never check `i = 1`. Besides, `iter_shape[0]` is now what was previously `iter_shape[1]`, so you never check it either.

Reply:
oh, wow! thanks for noticing this. it's weird that the tests pass ... will definitely require a fix.

(I have an implementation of all this logic in Python and did tests against all axis combinations vs numpy -- we could also use that implementation to make some nice animations and teach people how to do fast reductions in the ND case ;)